File: src/my_robot_tutorials/src/led_panel_node.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>
#include <my_robot_msgs/SetLed.h>


std::vector<bool> leds = {true, false, false};


// The callback of a server must return a boolean. True: Service Succesful
bool handle_set_led(
    my_robot_msgs::SetLed::Request &req,
    my_robot_msgs::SetLed::Response &res
)
{
    if (req.led_number < 1 || req.led_number > 3 || (req.state != 0 && req.state != 1)) {
        ROS_ERROR("Wrong SetLed request. LEDs stay unchanged");
        res.success = false;
        return true;
    }
    
    leds[req.led_number - 1] = (req.state == 1);
    ROS_WARN("Service called. Success");
    res.success = true;
    return true;
}


int main(int argc, char **argv)
{
    ros::init(argc, argv, "led_panel_node");
    ros::NodeHandle nh;

    ros::ServiceServer server = nh.advertiseService(
        "/set_led",
        handle_set_led
    );

    ros::Rate rate(1);

    ros::Publisher pub = nh.advertise<std_msgs::String>("/3_leds_status", 10);

    while(ros::ok()) {
        std_msgs::String msg;
        msg.data = "All LEDs status: LED 1: " + std::string(leds[0] ? "true" : "false") +
           ". LED 2: " + std::string(leds[1] ? "true" : "false") +
           ". LED 3: " + std::string(leds[2] ? "true" : "false");
        pub.publish(msg);

        ros::spinOnce(); // Handle callbacks

        rate.sleep();
    }
}

File: src/my_robot_tutorials/scripts/led_panel_node.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import String
from my_robot_msgs.srv import SetLed


leds = {
    1: True,
    2: False,
    3: False
}


def handle_set_led(req):
    global leds
    if req.led_number not in range(1, 4) or req.state not in [0, 1]:
        rospy.logerr("Wrong SetLed request. LEDs stay unchanged")
        return False
    
    leds[req.led_number] = req.state == 1
    rospy.logwarn("Service called. Success")
    return True


if __name__ == "__main__":
    # anonymous will add a random number after the name of the node. Therefore we can run multiple 
    # instances of this node
    rospy.init_node("led_panel")

    service = rospy.Service(
        "/set_led",  # Name of service
        SetLed,  # Type of request
        handle_set_led  # Handle function
    )
    rospy.loginfo("Service Server has been created")

    pub = rospy.Publisher("/3_leds_status", String, queue_size=10)

    rate = rospy.Rate(1)

    msg = String()
    while not rospy.is_shutdown():
        msg.data = "The Status of the LEDS is: " + str(leds)
        pub.publish(msg)
        rate.sleep()
